mlflow/pyfunc/__init__.py

```diff
@@ -1648,12 +1662,158 @@ def _convert_struct_values(
+     archive name because the prebuilt env might not be compatible across different
+     runtime version or platform machine.
+    """
+    python_env = _get_python_env(Path(local_model_path))
+    env_name = _get_virtualenv_name(python_env, local_model_path)
+    runtime_version, platform_machine = (
+        get_dbconnect_udf_sandbox_image_version_and_platform_machine(spark)
+    )
+    return f"{env_name}-{runtime_version}-{platform_machine}"
+
+
+def _verify_prebuilt_env(spark, local_model_path, env_archive_path):
+    from mlflow.utils.databricks_utils import (
+        get_dbconnect_udf_sandbox_image_version_and_platform_machine,
+    )
+
+    archive_name = os.path.basename(env_archive_path)[:-7]
+    prebuilt_env_sha, prebuilt_runtime_version, prebuilt_platform_machine = archive_name.split("-")[
+        -3:
+    ]
+
+    python_env = _get_python_env(Path(local_model_path))
+    env_sha = _get_virtualenv_name(python_env, local_model_path).split("-")[-1]
+    runtime_version, platform_machine = (
+        get_dbconnect_udf_sandbox_image_version_and_platform_machine(spark)
+    )
+
+    if prebuilt_env_sha != env_sha:
+        raise MlflowException(
+            f"The prebuilt env '{env_archive_path}' does not match the model required environment."
+        )
+    if prebuilt_runtime_version != runtime_version:
+        raise MlflowException(
+            f"The prebuilt env '{env_archive_path}' runtime version '{prebuilt_runtime_version}' "
+            f"does not match UDF sandbox runtime version {runtime_version}."
+        )
+    if prebuilt_runtime_version != runtime_version:
+        raise MlflowException(
+            f"The prebuilt env '{env_archive_path}' platform machine '{prebuilt_platform_machine}' "
+            f"does not match UDF sandbox platform machine {prebuilt_platform_machine}."
+        )
+
+
+def _prebuild_env_internal(local_model_path, archive_name, save_path):
+    env_root_dir = os.path.join(_PREBUILD_ENV_ROOT_LOCATION, archive_name)
+    archive_path = os.path.join(save_path, archive_name + ".tar.gz")
+    if os.path.exists(env_root_dir):
+        shutil.rmtree(env_root_dir)
+    if os.path.exists(archive_path):
+        os.remove(archive_path)
+
     try:
-        from pyspark.sql.utils import is_remote
-    except ImportError:
-        return False
-    return is_remote()
+        pyfunc_backend = get_flavor_backend(
+            local_model_path,
+            env_manager="virtualenv",
+            install_mlflow=False,
+            create_env_root_dir=False,
+            env_root_dir=env_root_dir,
+        )
+
+        pyfunc_backend.prepare_env(model_uri=local_model_path, capture_output=False)
+        # exclude pip cache from the archive file.
+        shutil.rmtree(os.path.join(env_root_dir, "pip_cache_pkgs"))
+
+        return archive_directory(env_root_dir, archive_path)
+    finally:
+        shutil.rmtree(env_root_dir, ignore_errors=True)
+
+
+def prebuild_model_env(model_uri, save_path):
```

Comment:
Is this a public API?

Reply:
`mlflow/pyfunc/__init__.py` is a public module.

Reply:
yes.

Reply:
then add param descriptions.

Reply:
param doc is added.

I'll add an example of "prebuild env + spark_udf" use case in a follow-up PR.

Reply:
can you add a small example in the docstring?

Reply:
added.